File: src/org/jgll/util/ToDotWithoutIntermeidateAndLists.java
package org.jgll.util;

import static org.jgll.util.GraphVizUtil.SYMBOL_NODE;

import org.jgll.sppf.ListSymbolNode;
import org.jgll.sppf.NonterminalSymbolNode;
import org.jgll.sppf.SPPFNode;
import org.jgll.traversal.SPPFVisitorUtil;

public class ToDotWithoutIntermeidateAndLists extends ToDotWithoutIntermediateNodes {
	
	@Override
	public void visit(NonterminalSymbolNode node) {
		SPPFVisitorUtil.removeIntermediateNode(node);
		
		if(!node.isVisited()) {
			node.setVisited(true);
	
			sb.append("\"" + getId(node) + "\"" + String.format(SYMBOL_NODE, replaceWhiteSpace(node.toString())) + "\n");
		
			for(SPPFNode child : node.getChildren()) {
				if(!child.getLabel().startsWith("layout(")) {
				  addEdgeToChild(node, child);
				  child.accept(this);
				}
			}
		}
	}
	
	@Override
	public void visit(ListSymbolNode node) {
		SPPFVisitorUtil.removeListSymbolNode(node);
		visit((NonterminalSymbolNode)node);
	}
	
}

File: src-scala/org/jgll/util/dot/ToDotWithoutIntermeidateAndLists.scala
package org.jgll.util.dot

import org.jgll.util.dot.GraphVizUtil.PACKED_NODE
import org.jgll.util.dot.GraphVizUtil.SYMBOL_NODE
import org.jgll.sppf.{ListSymbolNodeTrait, PackedNodeTrait, NonterminalSymbolNodeTrait}
import org.jgll.traversal.SPPFVisitorUtilTrait

trait ToDotWithoutIntermeidateAndListsTrait {
  self: ToDotWithoutIntermediateNodesTrait
   with NonterminalSymbolNodeTrait
   with SPPFVisitorUtilTrait
   with PackedNodeTrait
   with ListSymbolNodeTrait =>
  class ToDotWithoutIntermeidateAndLists extends ToDotWithoutIntermediateNodes {

    override def visit(node: NonterminalSymbolNode) {
      SPPFVisitorUtil.removeIntermediateNode(node)
      if (!node.isVisited) {
        node.setVisited(true)
        sb.append("\"" + getId(node) + "\"" +
          String.format(SYMBOL_NODE, replaceWhiteSpace(node.toString)) +
          "\n")
        for (child <- node.getChildren if !child.getLabel.startsWith("layout")) {
          addEdgeToChild(node, child)
          child.accept(this)
        }
      }
    }

    override def visit(node: PackedNode) {
      SPPFVisitorUtil.removeIntermediateNode(node)
      if (!node.isVisited) {
        node.setVisited(true)
        sb.append("\"" + getId(node) + "\"" + String.format(PACKED_NODE, "") +
          "\n")
        for (child <- node.getChildren if !child.getLabel.startsWith("layout")) {
          addEdgeToChild(node, child)
          child.accept(this)
        }
      }
    }

    override def visit(node: ListSymbolNode) {
      SPPFVisitorUtil.removeListSymbolNode(node)
      visit(node.asInstanceOf[NonterminalSymbolNode])
    }
  }
}